.github/workflows/pin-requirements.yml

```diff
@@ -0,0 +1,48 @@
+name: Pin Requirements
+
+on:
+  workflow_dispatch:
+  workflow_call:
+
+permissions:
+  pull-requests: write
+  contents: write
+
+jobs:
+  update-pinned-requirements:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@ac593985615ec2ede58e132d2e21d2b1cbd6127c # v3.3.0
+        with:
+          ref: main
+
+      - uses: actions/setup-python@d27e3f3d7c64b4bbf8e4abfb9b63b83e846e0435
+        with:
+          python-version: "3.x"
+          cache: "pip"
+          cache-dependency-path: pyproject.toml
+
+      - run: pip install pip-tools
+
+      - name: Determine latest sigstore-python PyPI release
+        run: |
+          version=$(curl -H "Accept: application/vnd.pypi.simple.v1+json" https://pypi.org/simple/sigstore/ | jq -r '.versions[-1]')
+          echo "sigstore_python_version=$version" >>$GITHUB_ENV
```

Comment:
Hmm, there might be some complexity here. When we run `pip-compile` on a branch's `pyproject.toml`, it doesn't include the hash of `sigstore` itself. Here are a few approaches that we can use to work around this:

- Rebuild the wheel in the`pin-requirements` workflow and pin with the built wheel
  - Is this approach reproducible? I tried my hand at building a wheel locally and seemed to generate two wheels with the same hash.
  - We'd have to duplicate some building logic from the `release` workflow
- Pass in the wheel from `release.yml` with a `workflow_call` input (what will we do with `workflow_dispatch`?)
- Don't check `sigstore`'s hash (probably not a good idea)

I'm leaning towards rebuilding the wheel in `pin-requirements`, since that still allows us to have a manual trigger with `workflow_dispatch`. I'm just not sure if the hash of the resultant wheel is guaranteed to be the same as the one we upload from the `release` workflow.

Reply:
Now that you mention CDN availability, the pins in the repo might be updated before new versions of `sigstore` are widely available. Are we comfortable with that propagation delay?

Reply:
> Is this approach guaranteed to be reproducible? I tried my hand at building a wheel locally and it seemed to generate two wheels with the same hash.

We don't guarantee reproducibility, so while it might work now, it might not work consistently.

Can't we just take the artifacts off the GitHub release itself?

Reply:
> > Is this approach guaranteed to be reproducible? I tried my hand at building a wheel locally and it seemed to generate two wheels with the same hash.
> 
> We don't guarantee reproducibility, so while it might work now, it might not work consistently.
> 
> Can't we just take the artifacts off the GitHub release itself?

Hmm, we should be able to do this. Do we want to keep the `workflow_dispatch` trigger? We'd be able to trigger manual updates on tags with associated releases.

Reply:
I'm not sure why it would generally be necessary, but might be good to keep around in case something fails. One of the inputs could just be the tag.